File: BFS/137.Clone Graph/Solution.py
"""
Definition for a undirected graph node
class UndirectedGraphNode:
    def __init__(self, x):
        self.label = x
        self.neighbors = []
"""


class Solution:
    """
    @param: node: A undirected graph node
    @return: A undirected graph node
    """
    def cloneGraph(self, node):
        # write your code here
        if node is None:
            return None

        # get all the nodes in the graph
        nodes = self._get_nodes(node)

        # map from original node to new node
        mapping = dict()

        # copy ndoes
        for graphNode in nodes:
            mapping[graphNode] = UndirectedGraphNode(graphNode.label)

        # copy edges
        for graphNode in nodes:
            for neighbor in graphNode.neighbors:
                newNeighbor = mapping[neighbor]
                mapping[graphNode].neighbors.append(newNeighbor)

        return mapping[node]


    def _get_nodes(self, node):
        from collections import deque

        queue = deque([])
        history = set()
        nodes = []

        queue.append(node)
        history.add(node)

        while queue:
            graphNode = queue.popleft()
            nodes.append(graphNode)

            for neighbor in graphNode.neighbors:
                if neighbor not in history:
                    queue.append(neighbor)
                    history.add(neighbor)

        return nodes

File: BFS/137.Clone Graph/Solution.java
/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     ArrayList<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */


public class Solution {
    /*
     * @param node: A undirected graph node
     * @return: A undirected graph node
     */
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        // write your code here
        if (node == null) {
            return null;
        }

        // Get all the nodes from input    
        List<UndirectedGraphNode> initialGraph = getGraphNodes(node);

        // Copy nodes
        HashMap<UndirectedGraphNode, UndirectedGraphNode> mapping = new HashMap<>();
        for (UndirectedGraphNode currNode : initialGraph) {
            mapping.put(currNode, new UndirectedGraphNode(currNode.label));
        }
        
        // Copy edges
        for (UndirectedGraphNode initialNode : initialGraph) {
            for (UndirectedGraphNode neighbor : initialNode.neighbors) {
                mapping.get(initialNode).neighbors.add(mapping.get(neighbor));
            }
        }

        return mapping.get(node);
    }

    private List<UndirectedGraphNode> getGraphNodes(UndirectedGraphNode node) {

        // BFS to get all nodes
        List<UndirectedGraphNode> graph = new ArrayList<>();
        Set<UndirectedGraphNode> history = new HashSet<>();
        Queue<UndirectedGraphNode> queue = new LinkedList<>();
        queue.offer(node);
        history.add(node);

        while (!queue.isEmpty()) {

            UndirectedGraphNode tmpNode = queue.poll();
            graph.add(tmpNode);

            for (UndirectedGraphNode neighbor : tmpNode.neighbors) {
                if (!history.contains(neighbor)) {
                    queue.offer(neighbor);
                    history.add(neighbor);
                }
            }

        }

        return graph;
    }

}